### ISSUE TITLE
Change exception raised from fileno() (and others) to io.UnsupportedOperation

### ISSUE THREAD

**Issue by metasyn** [pytest-dev/pytest#2276]
In capture, we define a method fileno for the DontReadFromInput:
https://github.com/pytest-dev/pytest/blob/master/_pytest/capture.py#L450

However, python docs suggest this shouldn't be implemented:
```
file.fileno()

Return the integer “file descriptor” that is used by the underlying implementation to request I/O operations from the operating system. This can be useful for other, lower level interfaces that use file descriptors, such as the fcntl module or os.read() and friends.

Note File-like objects which do not have a real file descriptor should not provide this method!
```

https://docs.python.org/2/library/stdtypes.html

Is there a particular reason it is provided?

**Comment by RonnyPfannschmidt** [comment]
@metasyn most likely to get better errors, but at close evaluation it looks a bit like its a bad idea for code thatd does checks - i wonder if there are other users that work that way

i checked `io.StringIO.fileno` and its implemented, so i believe its a good idea to check what is done, not what is said (and perhaps use a more specific exception

**Comment by nicoddemus** [comment]
Indeed:

```
Python 3.5.2 | packaged by conda-forge | (default, Jul 26 2016, 02:06:09) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import io
>>> io.StringIO().fileno()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
io.UnsupportedOperation: fileno
```

```
Python 2.7.11 (v2.7.11:6d1b6a68f775, Dec  5 2015, 20:32:19) [MSC v.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import io
>>> io.StringIO().fileno()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
io.UnsupportedOperation: fileno
```

(Same for 2.6)

I guess to follow suit we should raise `io.UnsupportedOperation` as well, instead of the current `ValueError`?

**Comment by RonnyPfannschmidt** [comment]
Yup

**Comment by RonnyPfannschmidt** [comment]
For compat we should check/ensure its a valueerror subclass

**Comment by nicoddemus** [comment]
According [to the docs](https://docs.python.org/2/library/io.html?highlight=unsupportedoperation#io.UnsupportedOperation):

> exception io.UnsupportedOperation
An exception inheriting OSError and ValueError that is raised when an unsupported operation is called on a stream.

So we are all set. 👍

**Comment by shubheksha** [comment]
I'd like to work on this! From what I understood by going through the above discussion is: I need to change the type of exception in `fileno` method from `ValueError` to `io.UnsupportedOperation`. Is that correct?

**Comment by RonnyPfannschmidt** [comment]
yup, in case of trouble on python2.6 we will assist

**Comment by metasyn** [comment]
Sorry slow response here - thanks for your thoughts ya'll. Sounds like a simple change. 
@shubheksha are you still interested in doing this? If not, I'd be happy to make a PR.

**Comment by RonnyPfannschmidt** [comment]
there is no PR so far, so just make one ^^

**Comment by metasyn** [comment]
@nicoddemus I summited an initial PR for this but realized you suggested in the title rename "(and others)" - could you let me know what you meant (or what I'm missing?)

Thanks!

**Comment by nicoddemus** [comment]
Hey thanks for the PR!

I was thinking about the other methods in that class which raise exceptions, specially `read` and `buffer` (but we must first see if those methods also raise the same exception).

**Comment by nicoddemus** [comment]
A quick look seems like the current behavior of `buffer` is correct:

```
>>> from io import StringIO
>>> StringIO().buffer
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: '_io.StringIO' object has no attribute 'buffer'
```

Not sure if we should change `read()` though.

**Comment by nicoddemus** [comment]
Resolved in #2313, thanks again @metasyn! 👍
